### PR TITLE
Make APIMethodMismatch exception verbose to print details of what actually happened

### LINKED ISSUE #9233: Fix ReqMgrAux postCampaignConfig API
When I invoke the following code:
```
from WMCore.Services.ReqMgrAux.ReqMgrAux import ReqMgrAux
mgr = ReqMgrAux(url, header=header, logger=logger)
mgr.postCampaignConfig(campaign, config)
```
it produces the following error:
```
httplib.HTTPException: url=https://vk-reqmgr.cern.ch/reqmgr2/data/data/campaignconfig/HINPbPbSpring18DR, code=405, reason=Method Not Allowed, headers={'Content-Length': '766', 'X-Error-Http': '405', 'Vary': 'Accept-Encoding', 'X-Rest-Time': '1501.083 us', 'Server': 'CherryPy/unknown', 'X-Error-Detail': 'API not supported for this request method', 'Allow': '', 'X-Rest-Status': '204', 'Date': 'Wed, 29 May 2019 20:07:14 GMT', 'Content-Type': 'text/html;charset=utf-8', 'CMS-Server-Time': 'D=20145 t=1559160434234088', 'X-Error-Id': '0c87d84b60cd9c769a19d4454f99cfb0'}
```
According to ReqMgr log the traceback shows that such method is not supported
```
[29/May/2019:22:07:14]  SERVER REST ERROR WMCore.REST.Error.APIMethodMismatch 0c87d84b60cd9c769a19d4454f99cfb0 (API not supported for this request method)
[29/May/2019:22:07:14]    Traceback (most recent call last):
[29/May/2019:22:07:14]      File "/data/srv/HG1904a/sw/slc7_amd64_gcc630/cms/reqmgr2/1.2.1.pre2/lib/python2.7/site-packages/WMCore/REST/Server.py", line 711, in default
[29/May/2019:22:07:14]        return self._call(RESTArgs(list(args), kwargs))
[29/May/2019:22:07:14]      File "/data/srv/HG1904a/sw/slc7_amd64_gcc630/cms/reqmgr2/1.2.1.pre2/lib/python2.7/site-packages/WMCore/REST/Server.py", line 760, in _call
[29/May/2019:22:07:14]        raise APIMethodMismatch()
[29/May/2019:22:07:14]    APIMethodMismatch: APIMethodMismatch 0c87d84b60cd9c769a19d4454f99cfb0 [HTTP 405, APP 204, MSG 'API not supported for this request method', INFO None, ERR None]
```

### PR DESCRIPTION
**Pull request by vkuznet**
Fixes https://github.com/dmwm/WMCore/issues/9233

#### Status
In development

#### Description
Provide details of failed API in exception message which will allow to understand what actually happens

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
https://github.com/dmwm/WMCore/issues/9233

#### External dependencies / deployment changes
no

### CODE REVIEW

**Comment by amaltaro** [comment]
@vkuznet can you paste here how the new exception would look like (including how/what API you called)?

**Review by amaltaro — APPROVED** [review]
Patch looks good! Can you just run a quick test and post the output as I requested before? Thanks

**Comment by vkuznet** [comment]
This is how error message looks on a client side:
```
httplib.HTTPException: url=https://vk-reqmgr.cern.ch/reqmgr2/data/data/campaignconfig/HINPbPbSpring18DR, code=405, reason=Method Not Allowed, headers={'Content-Length': '766', 'X-Error-Http': '405', 'Vary': 'Accept-Encoding', 'X-Rest-Time': '1338.959 us', 'Server': 'CherryPy/unknown', 'X-Error-Detail': 'API not supported for this request method', 'CMS-Server-Time': 'D=14756 t=1559561839540220', 'Allow': '', 'X-Rest-Status': '204', 'Date': 'Mon, 03 Jun 2019 11:37:19 GMT', 'Content-Type': 'text/html;charset=utf-8', 'X-Error-Info': 'Api "data" is not allowed, allowed methods:', 'X-Error-Id': 'cce9030d2a54079c5f17c27fd04ff7d5'}
```
and this is how it looks in a server log
```
APIMethodMismatch: APIMethodMismatch cce9030d2a54079c5f17c27fd04ff7d5 [HTTP 405, APP 204, MSG 'API not supported for this request method', INFO 'Api "data" is not allowed, allowed methods: ', ERR None]
```

So, now the error message clearly prints out which API (in quotes) is not allowed.

**Comment by vkuznet** [comment]
and, I called this url
https://vk-reqmgr.cern.ch/reqmgr2/data

SO, the wrong API is "data".

**Comment by amaltaro** [comment]
Valentin, this `APIMethodMistmatch` is meant to report an inconsistency between API and the HTTP method used, which wasn't your case. That's why in your examples the error message is incomplete (methods is empty).

To have a correct error message, we should actually create another exception class like `APINotSupported` or perhaps use the already existent `NoSuchInstance` class and then print the API called and the supported APIs under the same method.

If it's not clear to you, I can make another fix tomorrow.

**Comment by vkuznet** [comment]
Changed to APINotSupported, enjoy.

**Review by amaltaro — CHANGES_REQUESTED** [review]
Valentin, you're still missing one detail.
*) if the API can be found under a different method, then it should raise `APIMethodMismatch`
*) otherwise, it raises the new exception you created `APINotSupported`

The `msg` and `methods` variable might have to treated differently as well.

**Comment by vkuznet** [comment]
ok, how about now?

**Comment by amaltaro** [comment]
Almost there, but you can do better! The way you implemented it, I doubt it will ever raise `APIMethodMismatch`...

**Comment by vkuznet** [comment]
hmm, how about now?

File: src/python/WMCore/REST/Server.py [hunk 1]
```diff
@@ -767,9 +767,19 @@ def _call(self, param):
             raise APINotSpecified()
         api = param.args.pop(0)
         if api not in self.methods[request.method]:
-            response.headers['Allow'] = \
-                " ".join(sorted([m for m, d in self.methods.iteritems() if api in d]))
-            raise APIMethodMismatch()
+            methods = " ".join(sorted([m for m, d in self.methods.iteritems() if api in d]))
+            response.headers['Allow'] = methods
+            foundMethod = None
```
Comment (amaltaro) [inline src/python/WMCore/REST/Server.py]:
```suggestion
            if not methods:
                msg = 'Api "%s" not found. This method supports these Apis: %s' % (api, self.methods[request.method])
                raise APINotSupported(msg)
            else:
                msg = 'Api "%s" only supported in the method: "%s"' % (api, methods)
                raise APIMethodMismatch(msg)
```

how about the code above?

Comment (amaltaro) [reply src/python/WMCore/REST/Server.py]: well, and deleting a few lines below (first time I make this suggestion instead of a comment).

**Comment by vkuznet** [comment]
done

File: src/python/WMCore/REST/Server.py (hunk 1 above)
Comment (vkuznet) [reply src/python/WMCore/REST/Server.py]: done

**Comment by amaltaro** [comment]
test this please

**Comment by amaltaro** [comment]
Thanks, Valentin.

Here is how it looks now for `APIMethodMismatch`
```
$ curl --cert $X509_USER_CERT --key $X509_USER_KEY -X POST https://alancc7-cloud2.cern.ch/reqmgr2/data/info -H "Content-type: application/json" -d '{"UserDrainMode":true}' -vv
...
< HTTP/1.1 405 Method Not Allowed
< X-Error-Http: 405
< X-Error-Detail: API not supported for this request method
< X-Error-Info: Api "info" only supported in the method: "GET HEAD"
```

and for `APINotSupported`
```
$ curl --cert $X509_USER_CERT --key $X509_USER_KEY -X POST https://alancc7-cloud2.cern.ch/reqmgr2/data/infoooo -H "Content-type: application/json" -d '{"UserDrainMode":true}' -vv
...
< HTTP/1.1 404 Not Found
< X-Error-Http: 404
< X-Error-Detail: API not supported
< X-Error-Info: Api "infoooo" not found. This method supports these Apis: {'campaignconfig': {'validation': [<bound method CampaignConfig.validate of <WMCore.ReqMgr.Service.Auxiliary.CampaignConfig instance at 0x7...
```

Oups, this last line doesn't look good. We need to fetch only the keys `.keys()` and it would return the following info instead:
```
< X-Error-Info: Api "infoooo" not found. This method supports these Apis: ['campaignconfig', 'cmsswversions', 'request', 'splitting', 'wmagentconfig', 'unifiedconfig', 'permissions']
```

File: src/python/WMCore/REST/Server.py [hunk 2]
```diff
@@ -767,9 +767,14 @@ def _call(self, param):
             raise APINotSpecified()
         api = param.args.pop(0)
         if api not in self.methods[request.method]:
-            response.headers['Allow'] = \
-                " ".join(sorted([m for m, d in self.methods.iteritems() if api in d]))
-            raise APIMethodMismatch()
+            methods = " ".join(sorted([m for m, d in self.methods.iteritems() if api in d]))
+            response.headers['Allow'] = methods
+            if not methods:
+                msg = 'Api "%s" not found. This method supports these Apis: %s' % (api, self.methods[request.method])
```
Comment (amaltaro) [inline src/python/WMCore/REST/Server.py]:
```suggestion
                msg = 'Api "%s" not found. This method supports these Apis: %s' % (api, self.methods[request.method].keys())
```

Comment (amaltaro) [reply src/python/WMCore/REST/Server.py]: We're not in a hurry to get this in, so I'll let you commit and squash this single change.

Comment (vkuznet) [reply src/python/WMCore/REST/Server.py]: done

**Comment by vkuznet** [comment]
done

**Comment by amaltaro** [comment]
Thanks!

**Pull request merged**